src/CoreBundle/private/library/classes/TFullGroupTable/TFullGroupTable.class.php

```diff
@@ -415,6 +415,61 @@ private function getInlineFromAttributes(array $attributes)
         return $inline;
     }
 
+    private function getTableEditorPagedef($pagedef): string
+    {
+        $inputFilterUtil = $this->getInputFilterUtil();
+        $customTableEditor = $inputFilterUtil->getFilteredInput('sTableEditorPagdef');
+        if (null !== $customTableEditor && '' !== $customTableEditor) {
+            $pagedef = $customTableEditor;
+        }
+        return $pagedef;
+    }
+
+    private function getAjaxRecordsUrl(): string
+    {
+        $pagedef = $this->getTableEditorPagedef('tablemanager');
+        $inputFilterUtil = $this->getInputFilterUtil();
+        $tableId = $inputFilterUtil->getFilteredInput('id');
+        $urlUtil = ServiceLocator::get('chameleon_system_core.util.url');
+        $sAjaxURL = $urlUtil->getArrayAsUrl([
+            'id' => $tableId,
+            'pagedef' => $pagedef,
+            '_rmhist' => 'false',
+            'sOutputMode' => 'Ajax',
+            'module_fnc[contentmodule]' => 'ExecuteAjaxCall',
+            '_fnc' => 'getAutocompleteRecordList',
+            'sRestrictionField' => '',
+            'sRestriction' => '',
+            'recordID' => ''
+        ], PATH_CMS_CONTROLLER.'?', '&');
+
+        return $sAjaxURL;
+    }
+
+    private function switchToSelectedRecord(): string
+    {
+        $pagedef = $this->getTableEditorPagedef('tableeditor');
+        $inputFilterUtil = $this->getInputFilterUtil();
+        $tableId = $inputFilterUtil->getFilteredInput('id');
+        $urlUtil = ServiceLocator::get('chameleon_system_core.util.url');
```

Comment:
Could you move the ServiceLocator call to a separate method that includes a return type declaration? This improves type safety and keeps the distracting details of service retrieval away.

Reply:
Done.